### PR TITLE
change default value of verify to None

### LINKED ISSUE #3986: [BUG] change in 1.12.0 causes S3 SSLError
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [ ] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ x] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**:  macOS 10.15.6
- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: 1.12.0
- **Python version**: 3.8.5
- **npm version, if running the dev UI**:
- **Exact command to reproduce**:
S3ArtifactRepository(my_uri).list_artifacts(my_path)

### Describe the problem
Describe the problem clearly here. Include descriptions of the expected behavior and the actual behavior.

As of this version I'm getting an error:
"certificate verify failed: self signed certificate in certificate chain." 

I'm working with private corporate endpoints. This worked fine before 1.12.0. I've isolated it to changes in mlflow.store.artifact.s3_artifact_repo, line 54. Simply adding verify=True explicitly seems to have caused this, even though this should be the default in boto3. 

### Code to reproduce issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

Tricky 

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

### What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [x ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### LINKED ISSUE #4046: [BUG] MLFlow S3 artifact store configuration is not possible when using corporate proxy with resigned TLS certs
Thank you for submitting an issue. Please refer to our [issue policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md) for additional information about bug reports. For help with debugging your code, please refer to [Stack Overflow](https://stackoverflow.com/questions/tagged/mlflow).

**Please fill in this bug report template to ensure a timely and thorough response.**

### Willingness to contribute
The MLflow Community encourages bug fix contributions. Would you or another member of your organization be willing to contribute a fix for this bug to the MLflow code base?

- [X] Yes. I can contribute a fix for this bug independently.
- [ ] Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.
- [ ] No. I cannot contribute a bug fix at this time.

### System information
- **Have I written custom code (as opposed to using a stock example script provided in MLflow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: 

❯ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.1 LTS
Release:        20.04
Codename:       focal

- **MLflow installed from (source or binary)**: binary
- **MLflow version (run ``mlflow --version``)**: mlflow, version 1.13.1
- **Python version**: 3.7.9 (tested in 3.6 and 3.8 and had same issue)
- **npm version, if running the dev UI**: N/A
- **Exact command to reproduce**: The issue is only prevalent when using a corporate proxy with resigned TLS certs. Created a simple docker-compose deployment of MLFlow that reproduces the issue by using a self signed cert for Minio (representing S3): https://github.com/PeterSulcs/mlflow-issue 

### Describe the problem
MLFlow 1.11.0+ does not respect the setting of `REQUESTS_CA_BUNDLE` or `AWS_CA_BUNDLE` when an S3 artifact repo is used with a self signed cert.

### Code to reproduce issue
Provide a reproducible test case that is the bare minimum necessary to generate the problem.

```
git clone https://github.com/PeterSulcs/mlflow-issue.git
cd mlflow-issue
docker-compose up
```

Then, locally, with a Python 3.7.9 environment:

```
pip install mlflow==1.13.1 boto3 scikit-learn
source .env
python mlflow_test.py
```

### Other info / logs
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

```
python train.py
Traceback (most recent call last):
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connection.py", line 170, in _new_conn
    (self._dns_host, self.port), self.timeout, **extra_kw
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/util/connection.py", line 96, in create_connection
    raise err
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/util/connection.py", line 86, in create_connection
    sock.connect(sa)
ConnectionRefusedError: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connectionpool.py", line 706, in urlopen
    chunked=chunked,
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connectionpool.py", line 394, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connection.py", line 234, in request
    super(HTTPConnection, self).request(method, url, body=body, headers=headers)
  File "/home/psulcs/.pyenv/versions/3.7.9/lib/python3.7/http/client.py", line 1277, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/home/psulcs/.pyenv/versions/3.7.9/lib/python3.7/http/client.py", line 1323, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/home/psulcs/.pyenv/versions/3.7.9/lib/python3.7/http/client.py", line 1272, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/home/psulcs/.pyenv/versions/3.7.9/lib/python3.7/http/client.py", line 1032, in _send_output
    self.send(msg)
  File "/home/psulcs/.pyenv/versions/3.7.9/lib/python3.7/http/client.py", line 972, in send
    self.connect()
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connection.py", line 200, in connect
    conn = self._new_conn()
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connection.py", line 182, in _new_conn
    self, "Failed to establish a new connection: %s" % e
urllib3.exceptions.NewConnectionError: <urllib3.connection.HTTPConnection object at 0x7febcaae97d0>: Failed to establish a new connection: [Errno 111] Connection refused

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/connectionpool.py", line 756, in urlopen
    method, url, error=e, _pool=self, _stacktrace=sys.exc_info()[2]
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/urllib3/util/retry.py", line 573, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='127.0.0.1', port=5000): Max retries exceeded with url: /api/2.0/mlflow/experiments/get-by-name?experiment_name=example (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7febcaae97d0>: Failed to establish a new connection: [Errno 111] Connection refused'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "train.py", line 9, in <module>
    mlflow.set_experiment("example")
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/tracking/fluent.py", line 85, in set_experiment
    experiment = client.get_experiment_by_name(experiment_name)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/tracking/client.py", line 419, in get_experiment_by_name
    return self._tracking_client.get_experiment_by_name(name)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/tracking/_tracking_service/client.py", line 129, in get_experiment_by_name
    return self.store.get_experiment_by_name(name)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/store/tracking/rest_store.py", line 263, in get_experiment_by_name
    response_proto = self._call_endpoint(GetExperimentByName, req_body)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/store/tracking/rest_store.py", line 55, in _call_endpoint
    return call_endpoint(self.get_host_creds(), endpoint, method, json_body, response_proto)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/utils/rest_utils.py", line 160, in call_endpoint
    host_creds=host_creds, endpoint=endpoint, method=method, params=json_body
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/utils/rest_utils.py", line 79, in http_request
    max_rate_limit_interval, url=url, headers=headers, verify=verify, **kwargs
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/mlflow/utils/rest_utils.py", line 58, in request_with_ratelimit_retries
    response = requests.request(**kwargs)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/requests/api.py", line 61, in request
    return session.request(method=method, url=url, **kwargs)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/requests/sessions.py", line 542, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/requests/sessions.py", line 655, in send
    r = adapter.send(request, **kwargs)
  File "/home/psulcs/.pyenv/versions/mlflow-test/lib/python3.7/site-packages/requests/adapters.py", line 516, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPConnectionPool(host='127.0.0.1', port=5000): Max retries exceeded with url: /api/2.0/mlflow/experiments/get-by-name?experiment_name=example (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7febcaae97d0>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

[###](url) What component(s), interfaces, languages, and integrations does this bug affect?
Components 
- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by PeterSulcs**
## What changes are proposed in this pull request?

Change the default setting of `verify` arg to `None` rather than `true`. The default value of `true` isn't handled correctly by boto3, the library expects either `false`, `None` or a path to a cert bundle. This breaks the reference to the root cert that can be passed via the `REQUESTS_CA_BUNDLE` envar. This was a breaking change caused by PR that only affects cases where the artifact store has a certificate that will not be validated by the default root cert: https://github.com/mlflow/mlflow/pull/3345/files#diff-1fa0af6d2db749b300399663bbc25fcc867ab86a08b3e8c2b3c865561ddedf19

## How is this patch tested?

Patch was tested by running this example (https://github.com/PeterSulcs/mlflow-issue) with a dev install of this branch. Result is the same as mlflow 1.11.0.

### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [X] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

Closes #4046 and
Closes #3986 as pointed out below by @messlow

### CODE REVIEW

**Comment by APCG** [comment]
Will it be merged any time soon? We are also having this problem.

**Comment by APCG** [comment]
@andrewmchen, @dbczumar any feedback on this?

**Comment by PeterSulcs** [comment]
Anything I can do to help get this merged soon?

**Comment by PeterSulcs** [comment]
Bump. Is there anything I can do to help this get merged? 

This MR fixes an issue with S3 artifact configuration that causes MLFlow to pass an unexpected `verify` argument when creating the underlying boto3/botocore session.

Here is the section of the boto3 doc string explaining the acceptable values of the `verify` parameter.
https://github.com/boto/boto3/blob/73865126cad3938ca80a2f567a1c79cb248169a7/boto3/session.py#L212
```
        :type verify: boolean/string
        :param verify: Whether or not to verify SSL certificates.  By default
            SSL certificates are verified.  You can provide the following
            values:
            * False - do not validate SSL certificates.  SSL will still be
              used (unless use_ssl is False), but SSL certificates
              will not be verified.
            * path/to/cert/bundle.pem - A filename of the CA cert bundle to
              uses.  You can specify this argument if you want to use a
              different CA cert bundle than the one used by botocore.
```

By default, MLFlow currently passes `verify=True`, which sounds good, but this causes boto3 to silently ignore the `AWS_CA_BUNDLE` envar to verify against a custom CA bundle. Here is the section of MLFlow documentation that is relevant to the affected edge case: https://mlflow.org/docs/latest/tracking.html#amazon-s3-and-s3-compatible-storage 

 > If the MinIO server is configured with using SSL self-signed or signed using some internal-only CA certificate, you could set MLFLOW_S3_IGNORE_TLS or AWS_CA_BUNDLE variables (not both at the same time!) to disable certificate signature check, or add a custom CA bundle to perform this check, respectively:"


Digging deeper, this is what happens:

The `verify` argument of `True` gets passed to a botocore session:
https://github.com/boto/botocore/blob/develop/botocore/session.py#L801

Where **this** `if` statement evaluates to `False` instead of reading a configured certificate file from the ca_bundle envar because `True` is not `None`:
```
if verify is None:
            verify = self.get_config_variable('ca_bundle')
```
If the condition evaluates to true, the `get_config_variable` function then references the `AWS_CA_BUNDLE` envar. In current state, this `AWS_CA_BUNDLE` envar is never accessed, and therefore the only way to get mlflow communicate with the artifact repo is to set `MLFLOW_S3_IGNORE_TLS` to `true` (!!) or alter the certificate store that boto3/botocore reference(!!), both of which are non-ideal hacks.

TL;DR

I believe this MR is non-controversial. It corrects a small bug that causes MLFlow to violate the documented API to boto3. This bug is affecting anyone that is counting on MLFlow reading the `AWS_CA_BUNDLE` environment variable and using any version of MLFlow after 1.11.0. I have included an [example repo](https://github.com/PeterSulcs/mlflow-issue) that recreates the issue using a docker-compose file.

Please consider merging or letting me know what changes I can make! Thank you!

**Comment by Messlow** [comment]
Should we tag anyone with write access here to get traction for this bugfix?
It also related to this issue : https://github.com/mlflow/mlflow/issues/3986

**Comment by msharp9** [comment]
None of these failing tests seems related to the line changed in this MR.  Hoping we can get more eyes on this from those with write access.

**Comment by dolfinus** [comment]
I'm sorry that my PR #3345 caused such an issue :disappointed:. I was not aware of the fact that passing `True` option into boto3 library could cause such a behavior.

**Comment by APCG** [comment]
Any chance that this PR could be merged?

File: mlflow/store/artifact/s3_artifact_repo.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def _get_s3_client(self):
         s3_endpoint_url = os.environ.get("MLFLOW_S3_ENDPOINT_URL")
         ignore_tls = os.environ.get("MLFLOW_S3_IGNORE_TLS")
 
-        verify = True
+        verify = None
         if ignore_tls:
             verify = ignore_tls.lower() not in ["true", "yes", "1"]
```
Comment (WeichenXu123) [inline mlflow/store/artifact/s3_artifact_repo.py:47]: This line the verify will also be possible to set True.

shall we do:
```
        verify = True
        if ignore_tls:
            verify = ignore_tls.lower() not in ["true", "yes", "1"]

        if verify:
            verify = None # add comment why set None

```

**Review by WeichenXu123 — APPROVED** [review]
